steering_vectors/train_steering_vector.py

```diff
@@ -56,9 +56,18 @@ def train_steering_vector(
     layer_config = guess_and_enhance_layer_config(model, layer_config, layer_type)
     pos_activations: dict[int, list[Tensor]] = defaultdict(list)
     neg_activations: dict[int, list[Tensor]] = defaultdict(list)
+    n_pairs = len(training_samples)
+    if not isinstance(read_token_index, int):
+        pos_token_indices = read_token_index[:n_pairs]
+        neg_token_indices = read_token_index[n_pairs:]
```

Comment:
I'm a bit confused by this code. I assumed that if a list is passed for `read_token_index`, that would either mean that either:
-  a training pair is generated for every item in the list (e.g. we want a sample for both indices `-1` and `-2`).
- We somehow aggregate the activations from each item in the list to still get a single pos/neg pair for each training sample (e.g. we would take the mean of activations at indices `-1` and `-2`)

It looks like this is instead assuming that `read_token_index` is 2x the length of the training examples, and is setting a specific index for every example, where the first 0-n are for the first 0-n positive prompts, and n-2*n is for the corresponding negative prompts?

IMO if we want to still extract a single index for each prompt, but want to have the ability to set a different index per training sample, we shouldn't do it via passing in a 2x length list to the `read_token_index` param, since I feel this is hard to communicate to the user and easy to accidentally mess up. I would instead recommend adding something like `read_positive_token_index` and `read_negative_token_index` to `SteeringVectorTrainingSample`, and have the user pass in a `SteeringVectorTrainingSample` instance instead of a pos/neg prompt tuple. This way, it's clear that these values apply to the specific training sample they're attached to, and doesn't require communicating to the user that they need this 2x list length format. This is an advanced use-case, and for the basic use-case users could still use the tuple of `(pos, neg)` instead. If a `read_positive_token_index` or `read_negative_token_index` is provided for the `SteeringVectorTrainingSample`, that would override the `read_token_index` passed to the function.

So, it would look something like:

```python
@dataclass
class SteeringVectorTrainingSample:
    positive_prompt: str
    negative_prompt: str
    read_positive_token_index: Optional[int] = None
    read_negative_token_index: Optional[int] = None
```

What do you think of this proposal? Apologies if I'm misunderstanding the intention of this code!

Reply:
Another thought: Maybe the easiest way to achieve this is just to allow passing in a callable to `read_token_index` which takes in the prompt and returns an index. So, something like:

```python
read_token_index: int | Callable[[str], int] = -1
``` 

Then, the user can pass in a function to do whatever crazy logic they want, and it's pretty easy for us to implement and for the user to understand. For instance, if you wanted to find a specific token in the text, you could calculate that in the callable.

Reply:
> What do you think of this proposal?

You understood correctly and I agree with the concern that my current implementation isn't very intuitive to understand. I will instead add a Callable and extend `SteeringVectorTrainingSample` like you first suggested

Reply:
Done in https://github.com/steering-vectors/steering-vectors/pull/21/commits/76a9b9857d55a64d271626617708b33ad0b0e98e

Reply:
Nice, doing both also works!